server/src/main/java/org/elasticsearch/Version.java

```diff
@@ -192,7 +193,28 @@ public static Version fromId(int id) {
             case V_EMPTY_ID:
                 return V_EMPTY;
             default:
-                return new Version(id, org.apache.lucene.util.Version.LATEST);
+                // We need at least the major of the Lucene version to be correct.
+                // Our best guess is to use the same Lucene version as the previous
+                // version in the list, assuming that it didn't change. This is at
+                // least correct for patch versions of known minors since we never
+                // update the Lucene dependency for patch versions.
+                List<Version> versions = DeclaredVersionsHolder.DECLARED_VERSIONS;
+                Version tmp = new Version(id, org.apache.lucene.util.Version.LATEST);
+                int index = Collections.binarySearch(versions, tmp);
+                if (index < 0) {
```

Comment:
do we ever expect to find a match in the above collection if we end up in the default case here? Could this be `assert index < 0`?

Reply:
No indeed, I added an assertion.